source/extensions/filters/http/cache/cache_headers_utils.cc

```diff
@@ -168,6 +168,107 @@ absl::optional<uint64_t> CacheHeadersUtils::readAndRemoveLeadingDigits(absl::str
   return absl::nullopt;
 }
 
+absl::flat_hash_set<std::string> VaryHeader::parseAllowlist() {
+  // TODO(cbdm): Populate the hash_set from
+  // envoy::extensions::filters::http::cache::v3alpha::CacheConfig::allowed_vary_headers.
+  // Need to make sure that the headers we add here are valid values (i.e., not malformed). That
+  // way, we won't have to check this again in isAllowed.
+  return {"accept-encoding"};
+}
+
+bool VaryHeader::isAllowed(const Http::ResponseHeaderMap& headers) {
+  if (!hasVary(headers)) {
+    return true;
+  }
+
+  std::vector<std::string> varied_headers =
+      parseHeaderValue(headers.get(Http::Headers::get().Vary));
+
+  // If the vary value was malformed, it will not be contained in allowed_headers_.
+  for (const std::string& header : varied_headers) {
+    if (!allowed_headers_.contains(header)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool VaryHeader::hasVary(const Http::ResponseHeaderMap& headers) {
+  const Http::HeaderEntry* vary_header = headers.get(Http::Headers::get().Vary);
+  return ((vary_header) && (!vary_header->value().empty()));
+}
+
+std::string VaryHeader::createVaryKey(const Http::HeaderEntry* vary_header,
+                                      const std::vector<const Http::HeaderEntry*>& entry_headers) {
+  if (!vary_header) {
+    return "";
+  }
+
+  ASSERT(vary_header->key() == "vary");
+
+  std::string vary_key = "vary-key:";
+  std::string vary_key_values;
+
+  for (const std::string& header : parseHeaderValue(vary_header)) {
+    absl::StrAppend(&vary_key, header, ";");
+
+    for (const Http::HeaderEntry* cur_entry : entry_headers) {
+      if (cur_entry->key() == header) {
+        // TODO(cbdm): Can add some bucketing logic here based on header. For example, we could
+        // normalize the values for accept-language by making all of {en-CA, en-GB, en-US} into
+        // "en". This way we would not need to store multiple versions of the same payload, and any
+        // of those values would find the payload in the requested language.
+        absl::StrAppend(&vary_key_values, cur_entry->value().getStringView());
+        break;
+      }
+    }
+
+    absl::StrAppend(&vary_key_values, "\n");
+  }
+
+  absl::StrAppend(&vary_key, "\n", vary_key_values);
+
+  return vary_key;
+}
+
+std::vector<std::string> VaryHeader::parseHeaderValue(const Http::HeaderEntry* vary_header) {
```

Comment:
The problem is that we won't be able to use `StripAsciiWhitespace` and `AsciiStrToLower` if we use string_view.
They either take (i) a pointer to a string and change it in-place, or (ii) a string_view, create a new string and return it -- which defeats the purpose.

Reply:
Huh. I wouldn't've guessed that. Carry on then.

Reply:
I only figured it out after I did the change, also surprised me.

Reply:
See source/common/common/utility.h

It's easy to strip whitespace (`StringUtil::trim()`) without mutation (ie string_view -> string_view). If it were me I'd probably not lower-case it here, but instead use a case-insensitive compare on the hash lookup. Food for thought.

See `StringUtil::CaseInsensitiveHash` also.

I'm OK with what you have too, but it will churn a lot more memory during request processing, so maybe leave a TODO?

Reply:
Thanks for pointing out `StringUtil::trim`, I'll take a look at it.

Like the other discussion below, the allowlist from the config will be a StringMatcher.
So it won't be just a matter of using a case-insensitive hash.

I think we should make the header names lower-case, but I'm not sure.
What do you think?

Reply:
TBH I don't understand the implications of a vector of string matchers. I thought we might want to wind up with some kind of hash-set during operation; I just want to have that hash_set saved once rather than recomputed on every request.

I think we need a CacheFiilter Config. I don't need to have it in this PR but would ask for a TODO.

What I'm looking for in a filter is something that can be minimally allocated per-request without repeating a bunch of work that could be done just once.

Reply:
So because of the vector of string matchers, we would need to go through each one of them to check if the header name is allowed.
The matchers could be looking for exact, prefix, suffix, or regex matches, so I can't think how to generalize that to a hash-set, as-is.

From your comment, I think the minimal amount of parsing per request would mean saving the config as a member, and then just looping through the allowlist when checking if it's allowed.

Reply:
I'm not totally following your logic. What I'm saying is that there's a pattern in Envoy filters of computing a configuration object when config changes are pushed into Envoy, and generating filters from the config object.

There's also a shortcut pattern that generates filters directly from protos.

I think the cache should use the more complex pattern with an explicit config object. Once that's done we can compute things once and use them across many requests, such as creating a map or set from a vector of string matchers.

Reply:
Okay, so the CacheFilter does have a config proto. I think (if I'm understanding correctly) that if we use `allowed_vary_headers` from the config directly would do what you're suggesting: the allowlist would be used straight from config and not populated many times + it would be changed for a new config.

Is that on the right lines?

Reply:
Many filters have an explicit config object with a constructor. That allows you to convert the proto into a map or do other one-time processing to avoid doing it per-request.

In this case it's probably not necessary because usually the vary-list won't be long and a linear scan is OK. I suspect there may be other reasons you want to have an explicit config object rather than just referencing the xDS proto.

Reply:
CacheFilter handles config the way it does because I copied the best example I could find. Switching to a config object sounds like it could be a good idea, but it should be in a different PR.

Reply:
Some (maybe all) of the config will need to be able to be specified with greater granularity (per-route, if I understand Envoy routes correctly). Once I understand Envoy config mechanisms better, I expect to submit a PR to sort that out.

Reply:
Should we add a TODO or open a new issue as a reminder?

Reply:
ideally both, then you can write the TODO as `TODO(#9873987)` or whatever the number turns out to be.

Reply:
Please let me know if that's good or should I add more.